google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -671,8 +673,10 @@ private ApiFuture<WriteResult> executeWrite(
         processedOperationFuture,
         ApiException.class,
         e -> {
-          pendingOpsCount--;
-          processBufferedOperations();
+          synchronized (lock) {
+            pendingOpsCount--;
+            processBufferedOperations();
+          }
```

Comment:
Synchronization block to ensure that buffered operations are processed in expected order